examples/binarySearch/binarySearch.act

```diff
@@ -0,0 +1,45 @@
+behaviour init of BinarySearch
+interface constructor()
+
+creates
+	uint256[] array := []
+
+invariants
+	sorted(array)
+
+where
+
+sorted : uint256[] -> bool
+sorted([]) := true
+sorted([x]) := true
+sorted(x::y::xs) := x <= y and sorted(y::xs)
+
+
+// Returns the index of elem if elem in array,
+// otherwise returns -1.
+behaviour search of BinarySearch
+interface search(uint256 elem)
+
+returns binarySearch(array, elem, 0, array.length)
+
+ensures
+
+if elem in array
+then array[index] == elem
+else index == -1
+
+where
+
+index := binarySearch(array, elem, 0, array.length)
+
+binarySearch : uint256[] -> uint256 -> uint256 -> uint256 -> uint256
+binarySearch(uint256[] a, uint256 elem, uint256 left, uint256 right) :=
```

Comment:
What guarantees termination of this function?

Reply:
That would need to be proven by `something`

Reply:
I thought this is the spec. Spec should somehow be self sufficient. If you start putting code into spec things will get complicated.

Reply:
This example is just something you _could_ do if you want the reasoning to be done also inside the act, but don't necessarily have to. The motivation was the deposit contract, where we wanted the proof backends to try to prove its properties.